src/program/src/state.rs

```diff
@@ -1,511 +1,1244 @@
+        assert_eq!(
+            Err(TokenError::InvalidArgument),
+            State::process(&program_id, &mut account_infos, &instruction_data)
+        );
+
+        // create twice
+        let mut accounts = vec![
+            (&token_key, true, &mut token_account),
+            (&token_account_key, false, &mut token_account_account),
+        ];
+        let mut account_infos = create_is_signer_account_infos(&mut accounts);
+        assert_eq!(
+            Err(TokenError::InvalidArgument),
+            State::process(&program_id, &mut account_infos, &instruction_data)
+        );
+    }
 
-        let mut data = vec![0; 256];
+    #[test]
+    fn test_new_token_account() {
+        let program_id = new_pubkey(1);
+        let mut instruction_data = vec![0u8; size_of::<Command>()];
+        let token_account_key = new_pubkey(2);
+        let mut token_account_account = Account::new(0, size_of::<State>(), &program_id);
+        let owner_key = new_pubkey(3);
+        let mut owner_account = Account::default();
+        let token_key = new_pubkey(4);
+        let mut token_account = Account::new(0, size_of::<State>(), &program_id);
+
+        // missing signer
+        let instruction = Command::NewTokenAccount;
+        instruction.serialize(&mut instruction_data).unwrap();
+        let mut accounts = vec![
+            (&token_account_key, false, &mut token_account_account),
+            (&owner_key, false, &mut owner_account),
+            (&token_key, false, &mut token_account),
+        ];
+        let mut account_infos = create_is_signer_account_infos(&mut accounts);
+        assert_eq!(
+            Err(TokenError::MissingSigner),
+            State::process(&program_id, &mut account_infos, &instruction_data)
+        );
+
+        // create token account
+        let mut accounts = vec![
+            (&token_account_key, true, &mut token_account_account),
+            (&owner_key, false, &mut owner_account),
+            (&token_key, false, &mut token_account),
+        ];
+        let mut account_infos = create_is_signer_account_infos(&mut accounts);
+        State::process(&program_id, &mut account_infos, &instruction_data).unwrap();
+
+        // create twice
+        let mut accounts = vec![
+            (&token_account_key, true, &mut token_account_account),
+            (&owner_key, false, &mut owner_account),
+            (&token_key, false, &mut token_account),
+        ];
+        let mut account_infos = create_is_signer_account_infos(&mut accounts);
+        assert_eq!(
+            Err(TokenError::InvalidArgument),
+            State::process(&program_id, &mut account_infos, &instruction_data)
+        );
+    }
 
-        let account = TokenState::Account(TokenAccountInfo {
-            token: Pubkey::new(&[1; 32]),
-            owner: Pubkey::new(&[2; 32]),
-            amount: 123,
-            delegate: None,
+    #[test]
+    fn test_transfer() {
```

Comment:
nit: a bunch of test_transfer_xyz tests would be nicer to me than a single massive `test_transfer` test.  Anytime a unit test is longer than a page or so, it starts feeling unwieldy to me.

Reply:
Yeah, I don't like it much either but did so to avoid the duplicate code needed to setup each test.